test/src/util/pelz_json_parser_test_suite.c

```diff
@@ -0,0 +1,440 @@
+    cJSON_AddItemToObject(json_enc, "key_id_len", cJSON_CreateNumber(json_key_id_len));
+    cJSON_AddItemToObject(json_dec, "key_id", cJSON_CreateString(json_key_id[i]));
+    cJSON_AddItemToObject(json_dec, "key_id_len", cJSON_CreateNumber(json_key_id_len));
+    cJSON_AddItemToObject(json_enc, "enc_data", cJSON_CreateString(enc_data[i]));
+    cJSON_AddItemToObject(json_dec, "dec_data", cJSON_CreateString(dec_data[i]));
+    cJSON_AddItemToObject(json_enc, "enc_data_len", cJSON_CreateNumber(enc_data_len[i]));
+    cJSON_AddItemToObject(json_dec, "dec_data_len", cJSON_CreateNumber(dec_data_len[i]));
+
+    //Creating the request CharBuf for the JSON then testing request_decoder for encryption
+    tmp = cJSON_PrintUnformatted(json_enc);
+    request = newCharBuf(strlen(tmp));
+    memcpy(request.chars, tmp, request.len);
+    free(tmp);
+    CU_ASSERT(request_decoder(request, &request_type, &key_id, &data) == 0);
+    freeCharBuf(&request);
+    request_type = 0;
+    freeCharBuf(&key_id);
+    freeCharBuf(&data);
+
+    //Creating the request CharBuf for the JSON then testing request_decoder for decryption
+    tmp = cJSON_PrintUnformatted(json_dec);
+    request = newCharBuf(strlen(tmp));
+    memcpy(request.chars, tmp, request.len);
+    free(tmp);
+    CU_ASSERT(request_decoder(request, &request_type, &key_id, &data) == 0);
+    freeCharBuf(&request);
+    request_type = 0;
+    freeCharBuf(&key_id);
+    freeCharBuf(&data);
+
+    //Free the cJSON Objects to allow the addition of the next Object per the loop
+    cJSON_DeleteItemFromObject(json_dec, "dec_data");
+    cJSON_DeleteItemFromObject(json_dec, "dec_data_len");
+    cJSON_DeleteItemFromObject(json_enc, "enc_data");
+    cJSON_DeleteItemFromObject(json_enc, "enc_data_len");
+    cJSON_DeleteItemFromObject(json_enc, "key_id");
+    cJSON_DeleteItemFromObject(json_enc, "key_id_len");
+    cJSON_DeleteItemFromObject(json_dec, "key_id");
+    cJSON_DeleteItemFromObject(json_dec, "key_id_len");
+  }
+
+  cJSON_Delete(json_enc);
+  cJSON_Delete(json_dec);
+}
+
+void test_message_encoder(void)
+{
+  RequestType request_type[4] = { 0, 1, 2, 3 };
+  CharBuf key_id;
+  CharBuf data;
+  CharBuf message;
+  char *test[5] = { "file:/test/key1.txt", "test/key1.txt", "file", "anything", "" };
+
+  //Start Message Encoder Test
+  pelz_log(LOG_DEBUG, "Start Message Encoder Test");
+
+  data = newCharBuf(57);
+  memcpy(data.chars, "SwqqSZbNtN2SOfKGtE2jfklrcARSCZE9Tdl93pggkIsRkY3MrjevmQ==\n", data.len);
+  key_id = newCharBuf(strlen(test[0]));
+  memcpy(key_id.chars, test[0], key_id.len);
+  CU_ASSERT(message_encoder(request_type[0], key_id, data, &message) == 1);
+  CU_ASSERT(message_encoder(request_type[3], key_id, data, &message) == 1);
+  freeCharBuf(&key_id);
+
+  for (int i = 0; i < 5; i++)
+  {
+    key_id = newCharBuf(strlen(test[i]));
+    memcpy(key_id.chars, test[i], key_id.len);
+    CU_ASSERT(message_encoder(request_type[1], key_id, data, &message) == 0);
```

Comment:
Can we also check that the correct values are encoded? (I can see this being trickier than on the decode side, but if it's reasonably straightforward I think it would still be good to do.)